### PR TITLE
Feature/heatr

### PR DESCRIPTION
**Pull request by AmeliaJo**
first draft of heatr input checking

### CODE REVIEW

File: src/njoy21/input/HEATR/Card1/Nplot.hpp [hunk 1]
```diff
@@ -14,11 +14,13 @@ struct Nplot {
       "ascii text files. Negative values indicate unformatted input i.e.\n"
       "Fortran-style block binary\n"
       "\n"
-      "nplot values are restricted to an absolute value between 20 and 99,\n"
-      "inclusively, and are required to\n"
-      " + have an absolute value distinct from the corresponding nendf value\n"
-      " + have an absolute value distinct from the corresponding nin value\n"
-      " + have an absolute value distinct from the corresponding nout value";
+      "If the nplot option is used, then the nploy value are restricted to\n"
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card1/Nplot.hpp]: `If the nplot option is used, then the plot values are restricted to an absolute value`...

You should write here that `nplot` != 0 should only be used when `iprint`=2 from Card2

File: src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp [hunk 2]
```diff
@@ -1,103 +1,50 @@
...
-}
+SCENARIO( "nplot output values",
+  "[HEATR], [Card1], [Nplot]"){
+  GIVEN( "a valid nendf tape value" ){
+    Argument< HEATR::Card1::Nendf > nendf; nendf.value = 20;
+    Argument< HEATR::Card1::Nin   > nin;   nin.value   = 21;
+    Argument< HEATR::Card1::Nout  > nout;  nout.value  = 22;
+    
+    WHEN( "nplot is within range and not equal to nendf, nin, or nout" ){
+      std::vector<int> validValues{-23, 23, 0, 42, 99, -99};
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp]: You can't test the boundary like this. Probably won't matter, but it's like brown m&m's.

File: src/njoy21/input/HEATR/Card1/test/Card1.test.cpp [hunk 3]
```diff
@@ -1,41 +1,82 @@
...
-  }
-}
+SCENARIO( "Validating HEATR Card1 inputs",
+  "[HEATR], [Card1]"){
+  GIVEN( "valid HEATR Card1 inputs" ){
+    WHEN( "all inputs are provided" ){
+      iRecordStream<char> iss1( std::istringstream("20 21 22 23\n") );
+      iRecordStream<char> iss2( std::istringstream("-20 21 22 -23\n") );
+      iRecordStream<char> iss3( std::istringstream("20 -21 -22 23\n") );
+      THEN( "the returned values are correct" ){
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card1/test/Card1.test.cpp]: I like how this is organized.

File: src/njoy21/input/HEATR/Card2/Ed.hpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
+struct Ed {
+  using Value_t = Quantity<ElectronVolts>;
+  static std::string name(){ return "ed"; }
+  static std::string description(){
+    return
+      "ed is an optional integer argument that represents the damage\n"
+      "displacement energy, which is the energy (eV) required to displace an\n"
+      "atom from its lattice position.\n"
+      "\n"
+      "The default value for ed is defined by its corresponding element.\n"
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2/Ed.hpp]: Very, very nice.

File: src/njoy21/input/HEATR/Card2/Ed.hpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      "25 eV is used."; 
+  }
+
+  static Value_t defaultValue( Argument< HEATR::Card2::Matd >& matd ){ 
+    int mat = matd.value;
+    int za = 1000 * std::floor( mat / 100 );
+    int iz = std::round( za / 1000 );
+    auto default_val = 25.0 * electronVolts;
+
+    if      ( iz == 4  ){ default_val = 31.0 * electronVolts; }
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2/Ed.hpp]: Great work.

File: src/njoy21/input/HEATR/Card2/Ed.hpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      Log::warning( "Material specific default value not found in table\n"
+                    "default value of 25 eV used.\n" );
+    }
+    return default_val;
+
+  }
+
+  static bool verify( const Value_t& i,
+                      const Argument< HEATR::Card2::Matd >& ){ 
+    return (i > -1.0*electronVolts);
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2/Ed.hpp]: This still allows for a negative displacement energy. That doesn't sound right to me.

File: src/njoy21/input/HEATR/Card2/Ed/test/Ed.test.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            HEATR::Card2::Ed >( iss, matd ) );
+        }
+      } // THEN
+    } // WHEN
+  } // GIVEN
+  
+  GIVEN( "a card2 matd input value but no ed input value" ){
+    Argument< HEATR::Card2::Matd > matd;   
+    WHEN( "default value is in calculated table" ){
+      std::vector<int> validMatd{ 419, 600, 1225, 1325, 1425, 2025, 2225,
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2/Ed/test/Ed.test.cpp]: Very thorough.

File: src/njoy21/input/HEATR/Card2/Iprint/test/Iprint.test.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        for( int iprint : invalidValues ){
+          iRecordStream<char> iss( 
+            std::istringstream( std::to_string( iprint ) ) );
+          REQUIRE_THROWS( argument::extract< 
+            HEATR::Card2::Iprint >( iss, nplot ) );
+        }
+      } // THEN
+    } // WHEN
+  } // GIVEN
+  GIVEN( "a nonzero card1 nplot input value" ){
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2/Iprint/test/Iprint.test.cpp]: How come I can't give a thumbs up in a comment. Only when responding to a comment.

File: src/njoy21/input/HEATR/Card3/Mtk.hpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      "defined unless a gamma file for the mt is on the endf tape.\n"
+      "\n"
+      "The mt values should not be comma separated.";
+  }
+
+  static bool verify( const Value_t& mtk, 
+                      const Argument< HEATR::Card2::Npk >& ){
+    for( int mt : mtk ){
+      if( mt == 301 ){
+        Log::info( "mt value of 301 is automatically provided" );
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card3/Mtk.hpp]: Another thumbs up.

File: src/njoy21/input/HEATR/Card5/Qa.hpp [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      "or equal to 99 MeV requires a qbar (Card5a) entry.\n"
+      "\n"
+      "The q values should not be comma separated, and there should be\n"
+      "nqa entries in Card5 (nqa is an input from Card3).";
+  }
+
+  static bool verify( const Value_t& qa, 
+                      const Argument< HEATR::Card2::Nqa >& ){
+    for( auto q : qa ){
+      if( ( 99e6 * electronVolts <= q ) or ( -99e6 * electronVolts >= q ) ){
```
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: `if abs(q) > 99e6*electronVolt`

File: src/njoy21/input/HEATR.hpp [hunk 11]
```diff
@@ -1,11 +1,55 @@
...
-  // #include "njoy21/input/HEATR/Card3.hpp"
-  // #include "njoy21/input/HEATR/Card4.hpp"
-  // #include "njoy21/input/HEATR/Card5.hpp"
-  // #include "njoy21/input/HEATR/Card5a.hpp"
+  #include "njoy21/input/HEATR/Card3.hpp"
+  #include "njoy21/input/HEATR/Card4.hpp"
+  #include "njoy21/input/HEATR/Card5.hpp"
+  #include "njoy21/input/HEATR/Card5a.hpp"
 
+  using Card5aList = std::vector<Card5a>;
```
Comment (apmccartney) [inline src/njoy21/input/HEATR.hpp]: 👍

File: src/njoy21/input/HEATR.hpp [hunk 12]
```diff
@@ -1,11 +1,55 @@
...
+    if( card2.nqa.value > 0 ){
+      auto card4 = Card4( is, card2.nqa );
+      auto card5 = Card5( is, card2.nqa );
+      Card5aList card5aList;
+      for( const auto& qa : card5.qa.value ){
+        if( qa >= 99e6 * dimwits::electronVolts ){
+          card5aList.emplace_back( is );
+        }
+      }
+      ControlTuple ct( card4, card5, card5aList );
```
Comment (apmccartney) [inline src/njoy21/input/HEATR.hpp]: move rather than copy, especially when heap allocations are involved, e.g. vectors

Comment (apmccartney) [inline src/njoy21/input/HEATR.hpp]: In addition, prefer not to construct a value and return it. In this case, it precludes a [potential optimization](https://stackoverflow.com/a/12953129/2471708). Instead just return the temporary:
```cpp
return ControlTuple( std::move(card4), std::move(card5), std::move(Card5aList) );
```

File: src/njoy21/input/HEATR.hpp [hunk 13]
```diff
@@ -1,11 +1,55 @@
...
+      return ct;
+    }
+    
+    return std::nullopt;
+  }
+
+  template< typename Char >
+  HEATR( iRecordStream<Char>& is )
```
Comment (apmccartney) [inline src/njoy21/input/HEATR.hpp]: Why? 1 logical line of code.

File: src/njoy21/input/HEATR/Card2.hpp [hunk 14]
```diff
@@ -1,9 +1,56 @@
 class Card2 {
+  public:
   using Matd = argument::common::Matd;
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card2.hpp]: 👍

File: src/njoy21/input/HEATR/Card2/Ed.hpp (hunk 4 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card2/Ed.hpp]: Was impressed

File: src/njoy21/input/HEATR/Card2/Iprint/test/Iprint.test.cpp (hunk 8 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card2/Iprint/test/Iprint.test.cpp]: colon +1 colon

File: src/njoy21/input/HEATR/Card4/Mta.hpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+      "\n"
+      "The mt values should not be comma separated, and there should be\n"
+      "nqa entries in Card4 (nqa is an input from Card3).";
+  }
+
+  static bool verify( const Value_t& mta, 
+                      const Argument< HEATR::Card2::Nqa >& ){
+    for( int mt : mta ){
+      if( mt < 1 or mt > 999 ){
+        Log::info( "mt value {} is out of range", mt );
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card4/Mta.hpp]: It's best to echo the valid range of MT values in these sorts of messages

File: src/njoy21/input/HEATR/Card5/Qa.hpp (hunk 10 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: I'm not usre if `std::abs` is defined for quantities, but if not, it will be before the end of the day

File: src/njoy21/input/HEATR/Card5/Qa.hpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        Log::info( "\nq value: {}", q );
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: I actually would prefer this diagnostic removed.

File: src/njoy21/input/HEATR/Card5a/Qbar.hpp [hunk 17]
```diff
@@ -0,0 +1,17 @@
+struct Qbar {
+  using Value_t = ENDFtk::TAB1;
+  static std::string name(){ return "qbar"; }
+  static std::string description(){
+    return
+    "The qbar input is a endf format tab1 record giving qbar vs. energy.\n"
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5a/Qbar.hpp]: avoid math in english for documentation. rather than saying `qbar vs energy`, how about something like `energy released per reaction as a function of incident particle energy`

File: src/njoy21/input/HEATR/Card5a/Qbar.hpp [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    "qbar entries are used for Card5 qa values if a qa value has an\n"
+    "absolute value greater than 99 MeV.\n"
+    "\n"
+    "All x-values in the tab1 input must be greater than zero, and provided\n"
+    "in increasing order.";
+  }
+
+  static bool verify( const Value_t& qbar ){
+    return ( qbar.x().at( 0 ) > 0 );
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5a/Qbar.hpp]: this can throw which would be bad for verification. how about
```cpp
const auto energies = qbar.x();
return energies.size() and energies[0] > 0;
```

File: src/njoy21/input/HEATR/test/HEATR.test.cpp [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+	REQUIRE( card5aList.size() == 1 );
+	REQUIRE( card5aList[0].qbar.value.NR() == 1 );
+	REQUIRE( card5aList[0].qbar.value.NP() == 3 );
+	std::vector<double> x{8.9e6, 9.0e6, 1.1e7};
+	std::vector<double> y{-8.14e6, -8.14e6, -8.36e6};
+	REQUIRE( ranges::equal( x, card5aList[0].qbar.value.x() ) );
+	REQUIRE( ranges::equal( y, card5aList[0].qbar.value.y() ) );
+      } // THEN
+    } // WHEN
+
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/test/HEATR.test.cpp]: Please add a test with multiple card5a's (unless it's already there and I overlooked it)

File: src/njoy21/input/HEATR/Card1/Nout.hpp [hunk 20]
```diff
@@ -1,20 +1,9 @@
...
-      "ascii text files. Negative values indicate unformatted input i.e.\n"
-      "Fortran-style block binary\n"
-      "\n"
-      "nout values are restricted to an absolute value between 20 and 99,\n"
-      "inclusively, and are required to\n"
+      argument::common::Nout::description() +
+      "\n\n"
+      "For HEATR in particular, the nout input is is required to\n"
       " + have an absolute value distinct from the corresponding nendf value\n"
       " + have an value distinct from the corresponding nin value\n"
```
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card1/Nout.hpp]: also required to be distinct in absolute value from nin

**Comment by apmccartney** [comment]
Some responses to @jlconlin's, but that should be clear in the context. Good work Amelia.

File: src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp (hunk 2 above)
Comment (AmeliaJo) [inline src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp]: I'm a bit confused by this statement.. would it have been better to set nendf, nin, and nout to higher values so that nplot can be equal to 20? 
p.s. brown m&ms still deserve love

File: src/njoy21/input/HEATR/Card2/Ed.hpp (hunk 6 above)
Comment (AmeliaJo) [inline src/njoy21/input/HEATR/Card2/Ed.hpp]: Good catch thank you that was silly

File: src/njoy21/input/HEATR/Card5/Qa.hpp (hunk 10 above)
Comment (AmeliaJo) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: This doesn't seem to work because of its units

File: src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp (hunk 2 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp]: Correct

File: src/njoy21/input/HEATR/Card5/Qa.hpp (hunk 10 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: please file an issue on njoy21/dimensionalAnalysis. I'll fix that

File: src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp (hunk 2 above)
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp]: Yes, I was implying that `nendf`, `nin`, and `nout` should be larger so that we can check boundary conditions. See also [Brown M&Ms: A Quick Way to Determine Code Quality](http://hiltmon.com/blog/2015/10/14/brown-mms/)

File: src/njoy21/input/HEATR/Card5/Qa.hpp (hunk 10 above)
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: You can always do it for the underlying value while ignoring the unit. (I'm not sure what the syntax is for accessing the unit-less value.)

File: src/njoy21/input/HEATR/Card2.hpp (hunk 14 above)
Comment (jlconlin) [inline src/njoy21/input/HEATR/Card2.hpp]: How did you make a comment with a thumbs up? (I feel like an old man asking a teenager how to use SnapChat.

File: src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp (hunk 2 above)
Comment (AmeliaJo) [inline src/njoy21/input/HEATR/Card1/Nplot/test/Nplot.test.cpp]: This initial comment makes so much more sense now. Thank you for the link I was really confused about your candy reference

File: src/njoy21/input/HEATR/Card5/Qa.hpp (hunk 10 above)
Comment (apmccartney) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]:
```cpp
std::abs( q.value ) > 99e6
```

assuming `q` is a `Quantity<ElectronVolts>`

Comment (AmeliaJo) [inline src/njoy21/input/HEATR/Card5/Qa.hpp]: @apmccartney suggested I get rid of this check altogether (in further down comments). @jlconlin are you in agreement with that? I could see it reasonable to provide the warning or to remove it
